ldm/src/regex/regex.1
." $Id: regex.1,v 1.2.16.1.2.2 2009/06/18 16:19:38 steve Exp $
.TH REGEX 1 "$Date: 2009/06/18 16:19:38 $"
.SH NAME
regex \- Verify extended regular-expressions.
.SH SYNOPSIS
.HP
.ft B
regex
.nh
\%[-n\ \fIcount\fP]
\%[-s\ \fIstring\fP]
\%\fIpattern\fP]
.hy
.ft
.SH DESCRIPTION
.LP
This utility validates an extended regular-expression and, 
optionally, matches it against a specified string.
.LP
A looping option allows sufficient time to accumulate to permit the
determination
of the relative efficiency of, otherwise, equivalent regular-expressions.
.LP
Because a downstream LDM uses a extended regular-expression in its
request for data products from an upstream LDM, it is important that the
regular expression be efficient.  We have documented simple but 
pathological regular expressions from a downstream LDM request that have
caused the upstream LDM to spend over 99% of its user-time in the execution
of the regular expressions.  We're not making this up.
.SH OPTIONS
.TP
.BI "-n " count
The number of times to match the regular-expression against the string.
.TP
.BI "-s " string
The string to be matched against the regular-expression.  
.SH OPERAND
.TP
.I pattern
An extended reqular-expression to be compiled and, optionally, compared against
the argument of the
.B \-s
option.
.SH "STANDARD OUTPUT"
If the regular-expression is successfully compiled and the 
.B \-s
option is
used, then the utility prints "match" or "no match" as appropriate
to standard output.
.SH "STANDARD ERROR"
If the utility is incorrectly invoked or the regular-expression fails to
compile, then a diagnostic message is written to standard error.
.SH EXAMPLE
The regular-expression
.B [09]$
is MUCH faster than the regular-expression
.B .*[09]$
under Solaris:

.nf
.RS +4
$ time ./regex -n 10000 -s 'xxxxxxxxxxxxxxxxx' '[09]$'
no match

real    0m0.014s
user    0m0.010s    <-- NOTE
sys     0m0.010s
$ time ./regex -n 10000 -s 'xxxxxxxxxxxxxxxxx' '.*[09]$'
no match

real    0m1.870s
user    0m1.850s    <-- NOTE
sys     0m0.010s
.fi
.RE
.SH "EXIT STATUS"
.LP
The following exit values are returned:
.TP
0
Success.  Regular-expression successfuly compiled.
.TP
1
Usage error.
.TP
2
Regular-expression error.
.SH "SEE ALSO"
For information on extended regular-expressions, see
.B http://www.opengroup.org/onlinepubs/7908799/xbd/re.html.
.SH SUPPORT
.LP
If you have problems with this program, then you should first examine the 
LDM email archive for similar problems and how they were solved.
The email archive is available via the following World Wide Web URL:
.sp
.RS
\fBhttp://www.unidata.ucar.edu/software/ldm/\fP
.RE
.sp
If this does not suffice and your site is a member of the Unidata 
program, then send an inquiry via email -- together will all relevant 
information -- to
.sp
.RS
\fBsupport@unidata.ucar.edu\fP
.RE


ldm/src/regex/regex.c
/*
 *   Copyright 2003, University Corporation for Atmospheric Research
 *   See ../COPYRIGHT file for copying and redistribution conditions.
 */
/* $Id: regex.c,v 1.2.18.1 2005/01/21 21:35:15 steve Exp $ */


#include "ldmconfig.h"

#include <sys/types.h>  /* necessary for "off_t" under FreeBSD 4.7-STABLE */
#include <regex.h>
#include <stdio.h>
#include <unistd.h>


/*
 * This utility is useful for validating extended regular expressions used by
 * the LDM package.  In particular, a looping facility permits the determination
 * of the relative efficiency of, otherwise, equivalent regular-expressions.
 * For example, "[09]$" is MUCH faster than ".*[09]$" under Solaris:
 *
 * $ time ./regex -n 10000 -s 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' '.*[09]$'
 * no match
 * 
 * real    0m7.186s
 * user    0m7.160s    <-- NOTE
 * sys     0m0.010s
 * $ time ./regex -n 10000 -s 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' '[09]$'
 * no match
 * 
 * real    0m0.016s
 * user    0m0.000s    <-- NOTE
 * sys     0m0.010s
 *
 * Usage:
 *   regex [-n count] [-s string] pattern
 *
 *   where:
 *     count   - Number of times to match the regular-expression against the 
 *               string.  Default is 1.
 *     string  - The string to be matched against the regular-expression.  
 *     pattern - The regular-expression to be matched against the string.
 *
 * If no string is specified, then the regular-expression is simply compiled and
 * not matched against anything.
 *
 * If a string is specified and the regular expression is successfully compiled,
 * then the utility prints "match" or "no match" as appropriate to standard
 * output.
 *
 * Exit codes:
 *   0 - Success
 *   1 - Invalid usage.
 *   2 - Regular-expression failure.
 */
int main(const int argc, char* const argv[])
{
    int          retCode = 0;                   /* return code */
    int          ch;                            /* option character */
    int          count = 1;                     /* loop count */
    const char*  string = NULL;                 /* string to be matched */
    const char*  pattern;                       /* regular-expression */

    /*
     * Decode options.
     */
    while ((ch = getopt(argc, argv, "n:s:")) != -1) {
        switch (ch) {
            case 'n': {
                int i = sscanf(optarg, "%d", &count);

                if ((1 != i) || (1 > count)) {
                    fprintf(stderr, "Invalid \"-n\" argument: \"%s\"\n",
                        optarg);

                    retCode = 1;                /* usage error */
                }
                break;
            }

            case 's':
                string = optarg;
                break;
        }
    }

    /*
     * Obtain regular-expression string.
     */
    if (argc - 1 == optind) {
        pattern = argv[optind];
    }
    else {
        retCode = 1;                            /* usage error */
    }

    if (retCode) {
        fprintf(stderr, "Usage: %s [-n count] [-s string] pattern\n", argv[0]);
    }
    else {
        char    regMsg[512];                    /* RE error-message buffer */
        regex_t regExp;                         /* compiled RE */
        int     regCode;                        /* RE status */

        /*
         * Compile regular-expression string.
         */
        regCode = regcomp(&regExp, pattern, REG_EXTENDED);

        if (regCode != 0) {
            (void)regerror(regCode, &regExp, regMsg, sizeof(regMsg));
            (void)fprintf(stderr, "regcomp() failure: %s\n", regMsg);

            retCode = 2;                        /* recomp() failure */
        }
        else if (NULL != string) {
            int i;
            int matches;

            /*
             * Match compiled regular-expression against string for specified
             * number of times.
             */
            for (i = 0; i < count; i++) {       /* count > 0 */
                regCode = regexec(&regExp, string, 0, NULL, 0);

                if (0 == regCode) {
                    matches = 1;
                }
                else if (REG_NOMATCH == regCode) {
                    matches = 0;
                }
                else {
                    retCode = 2;                /* regexec() failure */
                    break;
                }
            }

            if (!retCode) {
                (void)printf("%s\n", matches ? "match" : "no match");
            }
            else {
                (void)regerror(regCode, &regExp, regMsg, sizeof(regMsg));
                (void)fprintf(stderr, "regexec() failure: %s\n", regMsg);
            }

            regfree(&regExp);
        }
    }

    return retCode;
}
